### PR TITLE
Adding Metadata Operations sample

### PR DESCRIPTION
**Pull request by JimDaly**
This sample demonstrates how to perform a wide range of data operations related to metadata and solutions.

### CODE REVIEW

File: cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,2312 @@
...
+    public static class CDSWebApiServiceExtensions
+    {
+        /// <summary>
+        /// The strong consistency header.
+        /// Strong consistency offers a linearizability guarantee. Linearizability refers to serving requests concurrently.
+        /// The reads are guaranteed to return the most recent committed version of an item.
+        /// Users are always guaranteed to read the latest committed write.
+        /// Use of this header incurs a small overhead and is therefore not suited to high volume workloads.
+        /// </summary>
+        private static readonly Dictionary<string, List<string>> STRONGCONSISTENCYHEADER = new Dictionary<string, List<string>>()
```
Comment (kalavany) [inline cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:18]: Can we name it strongConsistencyHeader?

 private static readonly Dictionary<string, List<string>> strongConsistencyHeader = new Dictionary<string, List<string>>()

File: cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,2312 @@
...
+            SolutionComponentType componentType,
+            string solutionUniqueName,
+            bool addRequiredComponents,
+            bool? doNotIncludeSubComponents = null,
+            string includedComponentsSettingsValues = null)
+        {
+            Guid solutionComponentId;
+
+            if (!string.IsNullOrEmpty(solutionUniqueName))
+            {
```
Comment (kalavany) [inline cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:45]: It's recommended to throw the exception in if clause. Then there would be no else clause.

if (string.IsNullOrEmpty(solutionUniqueName))
            {
throw new Exception("AddSolutionComponent requires the solutionUniqueName parameter value.");
}

File: cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,2458 @@
...
+                    throw new Exception($"There was a problem adding the {componentType} as a component to the {solutionUniqueName} solution:{ex.Message}");
+                }
+            }
+            else
+            {
+                throw new Exception("AddSolutionComponent requires the solutionUniqueName parameter value.");
+            }
+            return solutionComponentId;
+        }
+
```
Comment (phecke) [inline cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:82]: The next 3 methods can each be reduced to one call to svc.Post() with a inline anonymous JObject definition without the need to code a method. This is more a matter of the author's writing style and is just mentioned here as an FYI.

File: cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,2458 @@
...
+            Guid solutionComponentId;
+
+            if (!string.IsNullOrEmpty(solutionUniqueName))
+            {
+                //Verify that solution exists and is unmanaged.
+                try
+                {
+                    solutionId = svc.CheckSolution(solutionUniqueName);
+                }
+                catch (Exception)
```
Comment (phecke) [inline cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:186]: Why catch an exception and do nothing more than throw it? Perhaps this was a placeholder for you to add some code. Anyway, this shows up in many places in your code. I would recommend removing it unless you are doing more than throwing the caught exception.

File: cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,2458 @@
...
+        /// <param name="lookup">The customer lookup to create.</param>
+        /// <param name="oneToManyRelationships">A pair of relationships to create for the account and contact entities.</param>
+        /// <param name="solutionUniqueName">The unique name of the unmanaged solution.</param>
+        /// <returns>Id values of the created attribute and relationships</returns>
+        public static CreateCustomerRelationshipsResponse CreateCustomerRelationships(this CDSWebApiService svc,
+            ComplexLookupAttributeMetadata lookup,
+            List<ComplexOneToManyRelationshipMetadata> oneToManyRelationships,
+            string solutionUniqueName = null)
+        {
+            if (!string.IsNullOrEmpty(solutionUniqueName))
```
Comment (phecke) [inline cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:235]: You are checking if the solution name is null or empty several times in this method. Best to check once at the beginning of the method, and also call CheckSolution(). Once that is done you are good to go for the rest of the method.

File: cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,2458 @@
...
+
+        /// <summary>
+        /// Deletes an entity by Id
+        /// </summary>
+        /// <param name="svc">An instance of the CDSWebApiService to extend.</param>
+        /// <param name="entityMetadataId">The MetadataId of the entity to delete.</param>
+        public static void DeleteEntity(this CDSWebApiService svc,
+            Guid entityMetadataId)
+        {
+            svc.DeleteEntity(entityMetadataId, null);
```
Comment (phecke) [inline cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:623]: For this method and the ones that follow, it is not recommended to have methods that do nothing other than call another method.

File: cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,2458 @@
...
+        public static bool CanBeReferenced(this CDSWebApiService svc, string entityName)
+        {
+            try
+            {
+                JObject body = new JObject()
+                {
+                    ["EntityName"] = entityName
+                };
+                return (bool)svc.Post("CanBeReferenced", body)["CanBeReferenced"];
+            }
```
Comment (phecke) [inline cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:98]: General observation that is replicated throughout this code file. You have try/catch blocks in which you create a new exception with the caught exception as the inner exception and the parent exception message contains the inner exception message. Or you create an exception that tells the end user what method the exception occurred in. Neither of these are needed as you will be able to find this info when using a debugger.

Perhaps you like having such info printed out in the console without having to use a debugger? The problem with this approach is that 1) our dev customers know and expect to use a debugger, and 2) if they want to copy and use the code in their own projects (which they do) they will have to strip this code out.

**Review by phecke — CHANGES_REQUESTED** [review]
I went through the first 1000 lines in the Extensions.cs file (see inline comments), plus my comments for the CDSWebApiService.cs file in email. That is all I had the time for right now.

File: cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs (hunk 1 above)
Comment (JimDaly) [reply cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs]: Done

File: cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs (hunk 2 above)
Comment (JimDaly) [reply cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:45]: Done

File: cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs (hunk 3 above)
Comment (JimDaly) [reply cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:81]: Thanks. 
While each of these have a similar pattern, the goal is to show how to use the named actions rather than write a minimal amount of code.
[CanBeReferenced Action](https://docs.microsoft.com/en-us/dynamics365/customer-engagement/web-api/canbereferenced?view=dynamics-ce-odata-9)
[CanBeReferencing Action](https://docs.microsoft.com/en-us/dynamics365/customer-engagement/web-api/canbereferencing?view=dynamics-ce-odata-9)
[CanManyToMany Action](https://docs.microsoft.com/en-us/dynamics365/customer-engagement/web-api/canmanytomany?view=dynamics-ce-odata-9)

File: cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs (hunk 4 above)
Comment (JimDaly) [reply cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:185]: This code is calling another method `CheckSolution ` which throws a detailed exception.
This is just verifying that the solution they named is valid before proceeding. 
Yes, it is called this way multiple times because it provides the same solution checking each time the solutionUniqueName parameter is passed.

File: cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs (hunk 5 above)
Comment (JimDaly) [reply cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:234]: This operation is to create a customer relationship and solutionUniqueName is an optional parameter which, if included, will associate the 3 created solution components with the unmanaged solution indicated by the solutionUniqueName.

- First check is to ensure that the solution exists and is managed.
- Second check is to only include the value in the JObject body when it exists
- Final check is after the 3 solution components are created, then they can be associated with the named solution if it exists.

In other words, the association with a solution is a secondary, optional operation within the flow. It doesn't control the main flow, just supplements it when provided. Therefore, it seems appropriate to use the existence of the parameter when needed (3 times) to control the secondary optimal options.

File: cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs (hunk 6 above)
Comment (JimDaly) [reply cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:622]: There are two mutually exclusive ways to access many metadata items: by Metadata Id or by Name which is an alternate key.
See: [Retrieve metadata by name or MetadataId](https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/webapi/retrieve-metadata-name-metadataid)

These methods do more than just call another method, they provide overrides that can be easily used depending on the developer's situation. One accepts a Guid id value, the other a string name value. Both of them call the same private method that can accept both and manage passing the correct one.

It is intentional and a means to simplify the developer experience while re-using the same underlying logic.

File: cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs (hunk 7 above)
Comment (JimDaly) [reply cds/webapi/C#/MetadataOperations/CDSWebApiServiceExtensions.cs:97]: I'd be happy to discuss alternative patterns. The error handling I've done is focused on identifying the source of the error while debugging the console application sample. This is somewhere in-between the raw-low level samples and an SDK.

If you have a different pattern that you would recommend, please submit a PR and I can review it.

**Comment by JimDaly** [comment]
@phecke 
Do you have any specific changes you want made?
I think I've addressed your comments and I responded to your questions in email.
I don't see anything that has to block this PR so if I don't hear back from you in a couple days I will go ahead and merge.

**Comment by phecke** [comment]
@JimDaly My comments were suggestions. No significant problems identified. Go ahead and merge the PR.
